forecasttools/idata_to_tidy.py

```diff
@@ -0,0 +1,84 @@
+"""
+Contains functions for interfacing between
+the tidy-verse and arviz, which includes
+the conversion of idata objects (and hence
+their groups) in tidy-usable objects.
+"""
+
+import arviz as az
+import polars as pl
+import polars.selectors as cs
+
+
+def convert_inference_data_to_tidydraws(
+    idata: az.InferenceData, groups: list[str]
+) -> dict[str, pl.DataFrame]:
+    """
+    Creates a dictionary of polars dataframes
+    from the groups of an arviz InferenceData
+    object for use with the tidybayes API.
+
+    Parameters
+    ----------
+    idata : az.InferenceData
+        An InferenceData object.
+    groups : list[str]
+        A list of groups to transform to
+        tidy draws format. Defaults to all
+        groups in the InferenceData.
+
+    Returns
+    -------
+    dict[str, pl.DataFrame]
+        A dictionary of groups from the idata
+        for use with the tidybayes API.
+    """
+    available_groups = list(idata.groups())
+    if groups is None:
+        groups = available_groups
+    else:
+        invalid_groups = [
+            group for group in groups if group not in available_groups
+        ]
+        if invalid_groups:
+            raise ValueError(
+                f"Requested groups {invalid_groups} not found"
+                " in this InferenceData object."
+                f" Available groups: {available_groups}"
+            )
+
+    idata_df = pl.DataFrame(idata.to_dataframe())
+
+    tidy_dfs = {
+        group: (
+            idata_df.select("chain", "draw", cs.starts_with(f"('{group}',"))
+            .rename(
+                {
+                    col: col.split(", ")[1].strip("')")
+                    for col in idata_df.columns
+                    if col.startswith(f"('{group}',")
+                }
+            )
+            # draw in arviz is iteration in tidybayes
+            .rename({"draw": ".iteration", "chain": ".chain"})
+            .unpivot(
+                index=[".chain", ".iteration"],
+                variable_name="variable",
+                value_name="value",
+            )
+            .with_columns(
+                pl.col("variable").str.replace(r"\[.*\]", "").alias("variable")
+            )
+            .with_columns(pl.col(".iteration") + 1, pl.col(".chain") + 1)
+            .pivot(
+                values="value",
+                index=[".chain", ".iteration"],
+                columns="variable",
+                aggregate_function="first",
```

Comment:
should be `None` [per the docs](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.pivot.html), no? (unless I'm misunderstanding what your goal is with this operation). You want one column for each unique value of `"variable"` for a given `".chain"` and `".iteration"`.

```suggestion
                aggregate_function=None,
```

https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.pivot.html

Reply:
For the actual `pyrenew` inference data object, but not for the other examples I have made, using `None` and not `first` produces this error:

> FAILED tests/test_idata_to_tidy.py::test_posterior_predictive_group - polars.exceptions.ComputeError: found multiple elements in the same group, please specify an aggregation function

Reply:
That suggests something is not correct upstream. Perhaps the variable name regex?

Reply:
I believe the source of this issue is the line in `col: col.split(", ")[1].strip("')")`

```
.rename(
    {
        col: col.split(", ")[1].strip("')")
        for col in idata_df.columns
        if col.startswith(f"('{group}',")
    }
)
```

Reply:
I changed

`col: col.split(", ")[1].strip("')")`

to

`col: re.search(r",\s*'?(.+?)'?\)", col).group(1)`

which I got incorrect a handful of times but which I believe works now. 

The target of this regex looks like, e.g.,  `"('posterior', 'alpha')"`. 

In the expression, the `,` matches comma; in `\s*`, the `\s` is spaces and `*` for zero or more spaces; in `'?`, the `'` means single quote and `?` means optional (0 or 1 times); in `(.+?)`, the `(` and `)` capture whatever is in the parenthesis with `.` means non-newline character, `+` means one or more characters, and `?` (which I am not sure is necessary to include here) tries to get this capture in as small a job as possible (this capture will get `alpha` in the example target); the `'?` is the same as before; the `\)` is the single closing parenthesis after the group (the `\` is needed for escaping parenthesis; and `.group(1)` in `re` gets the first item captured by `(.+?)`.

Reply:
Can you write a test to demonstrate the case where the previous version was failing and the new version succeeds? They both work correctly on the example provided in your comment: `"('posterior', 'alpha')"`.

Reply:
The split approach did not work on the InferenceData object from `pyrenew-hew` in tests when the aggregate function was not set to `first`. Yes, both the split and `re` approaches work on "('posterior', 'alpha')".

Reply:
Thank you for comment + I will write test.

Reply:
> The split approach did not work on the InferenceData object from pyrenew-hew in tests when the aggregate function was set to None but did work when the aggregate function was set to first. Yes, both the split and re approaches work on "('posterior', 'alpha')".

I should probably investigate why, exactly, aggregate=None w/ split doesn't work for the Pyrenew inference data.